### PR TITLE
Make `RepositoryCache#put` resistant to concurrent file system operations

### PR DESCRIPTION
**Pull request by fmeum**
This requires making the implementation of `renameTo` in `JavaIoFileSystem` and `WindowsFileSystem` atomic.

Fixes https://github.com/bazelbuild/bazel/issues/21823

### CODE REVIEW

File: src/test/shell/bazel/starlark_repository_test.sh [hunk 1]
```diff
@@ -3183,4 +3183,97 @@ EOF
...
+  assert_equals 1 "$(wc -l < bazel-bin/unique_hashes)"
+  assert_equals $sha "$(cat bazel-bin/unique_hashes)"
+
+  # Verify that the repository cache entry has been created.
+  cache_entry="$repo_cache_dir/content_addressable/sha256/$sha/file"
+  ls -R "$repo_cache_dir" >& $TEST_log
+  echo "$sha $cache_entry" | sha256sum --check || fail "sha256 mismatch"
+}
+
+function test_repository_cache_concurrency_with_integrity() {
```
Comment (meteorcloudy) [inline src/test/shell/bazel/starlark_repository_test.sh:3233]: what's the difference from the previous test case?

**Review by meteorcloudy — COMMENTED** [review]
Nice, not sure what the test failure on macOS arm64 means?
```
test_repository_cache_concurrency FAILED: Expected '1', was '       1'.
```

File: src/test/shell/bazel/starlark_repository_test.sh (hunk 1 above)
Comment (fmeum) [reply src/test/shell/bazel/starlark_repository_test.sh]: This one specifies a `sha256`. But I don't think the other one provides any additional coverage, so I removed it.

**Review by Wyverald — APPROVED** [review]

**Comment by fmeum** [comment]
@bazel-io fork 7.2.0

**Pull request closed without merging**